airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/HeartbeatTimeoutChaperone.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers.internal;
+
+import static java.lang.Thread.sleep;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.featureflag.FeatureFlagClient;
+import io.airbyte.featureflag.ShouldFailSyncIfHeartbeatFailure;
+import io.airbyte.featureflag.Workspace;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link HeartbeatTimeoutChaperone} takes in an arbitrary runnable and a heartbeat monitor. It
+ * runs each in separate threads. If the heartbeat monitor thread completes before the runnable,
+ * that means that the heartbeat has stopped. If this occurs the chaperone cancels the runnable
+ * thread and then throws an exception. If the runnable thread completes first, the chaperone
+ * cancels the heartbeat and then returns.
+ *
+ * This allows us to run an arbitrary runnable that we can kill if a heartbeat stops. This is useful
+ * in cases like the platform reading from the source. The thread that reads from the source is
+ * allowed to run as long as the heartbeat from the sources is fresh.
+ */
+public class HeartbeatTimeoutChaperone {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HeartbeatTimeoutChaperone.class);
+
+  public static final Duration DEFAULT_TIMEOUT_CHECK_DURATION = Duration.ofMinutes(1);
+
+  private final HeartbeatMonitor heartbeatMonitor;
+  private final Duration timeoutCheckDuration;
+  private final FeatureFlagClient featureFlagClient;
+  private final UUID workspaceId;
+  private final ExecutorService executorService = Executors.newFixedThreadPool(2);
+  private final Optional<Runnable> customMonitor;
+
+  private boolean hasFailed = false;
+
+  public HeartbeatTimeoutChaperone(final HeartbeatMonitor heartbeatMonitor,
+                                   final Duration timeoutCheckDuration,
+                                   final FeatureFlagClient featureFlagClient,
+                                   final UUID workspaceId) {
+    this.timeoutCheckDuration = timeoutCheckDuration;
+
+    this.heartbeatMonitor = heartbeatMonitor;
+    this.featureFlagClient = featureFlagClient;
+    this.workspaceId = workspaceId;
+    this.customMonitor = Optional.empty();
+  }
+
+  @VisibleForTesting
```

Comment:
Nit:  This annotation can be removed, as the constructor is public.

Reply:
I move the ctor to default scope instead.